src/idl/src/directive.c

```diff
@@ -101,150 +109,168 @@ push_source(idl_pstate_t *pstate, const char *inc, const char *abs, bool sys)
   return IDL_RETCODE_OK;
 }
 
-#define START_OF_FILE (1u<<0)
-#define RETURN_TO_FILE (1u<<1)
-#define SYSTEM_FILE (1u<<2)
-#define EXTRA_TOKENS (1u<<3)
-
 static void delete_line(void *ptr)
 {
   struct line *dir = (struct line *)ptr;
   assert(dir);
+  if (dir->path)
+    free(dir->path);
   if (dir->file)
     free(dir->file);
   free(dir);
 }
 
-static idl_retcode_t
-push_line(idl_pstate_t *pstate, struct line *dir)
+static idl_retcode_t push_line(idl_pstate_t *pstate, struct line *dir)
 {
-  idl_retcode_t ret = IDL_RETCODE_OK;
+  idl_retcode_t ret;
 
-  assert(dir);
   if (dir->flags & (START_OF_FILE|RETURN_TO_FILE)) {
-    bool sys = (dir->flags & SYSTEM_FILE) != 0;
-    char *norm = NULL, *abs, *inc;
-    abs = inc = dir->file;
-    /* convert to normalized file name */
-    if (!idl_isabsolute(abs)) {
-      /* include paths are relative to the current file. so, strip file name,
-         postfix with "/relative/path/to/file" and normalize */
-      const char *cwd = pstate->scanner.position.source->path->name;
-      const char *sep = cwd;
-      assert(idl_isabsolute(cwd));
-      for (size_t i=0; cwd[i]; i++) {
-        if (idl_isseparator(cwd[i]))
-          sep = cwd + i;
+    char *norm = NULL;
+    const idl_source_t *src = pstate->scanner.position.source;
+
+    if (!idl_isabsolute(dir->path)) {
+      const char *cwd = src->path->name;
+
+      if (cwd && strcmp(cwd, "<builtin>") != 0) {
+        char *abs = NULL;
+        size_t len = 0, sep = 0;
+        assert(idl_isabsolute(cwd));
+        for (size_t pos=0; cwd[pos]; pos++) {
+          if (!idl_isseparator(cwd[pos]))
+            sep = 0;
+          else if (!sep)
+            len = sep = pos;
+        }
+        if (idl_asprintf(&abs, "%.*s", len, cwd, dir->path) < 0)
```

Comment:
I left it in for now. But yes, it's a bit of a left-over from thinking we might have to signal a system include and since we switched compiler types, that won't ever happen.